rtgui/editcallbacks.h

```diff
@@ -0,0 +1,256 @@
+/*
+ *  This file is part of RawTherapee.
+ *
+ *  Copyright (c) 2019 Jean-Christophe FRISCH <natureh.510@gmail.com>
+ *
+ *  RawTherapee is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  RawTherapee is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with RawTherapee.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "editid.h"
+#include "cursormanager.h"
+#include "../rtengine/coord.h"
+
+class Geometry;
+class EditDataProvider;
+
+/** @file
+ * See editwidgets.h for documentation
+ */
+
+/// @brief Method for client tools needing Edit information
+class EditSubscriber
+{
+
+public:
+
+private:
+    EditUniqueID ID; /// this will be used in improcfun to locate the data that has to be stored in the buffer; it must be unique in RT
+    EditType editingType;
+    BufferType bufferType;
+    EditDataProvider *provider;
+
+protected:
+    std::vector<Geometry*> visibleGeometry;    /// displayed geometry

```

Comment:
Again, who owns `Geometry*`? Maybe a `std::unique_ptr<>` is due.

Reply:
@Floessie At first sight, yes, it should make things simpler to use a `std::unique_ptr<>`. I'll look at this.

Reply:
@Floessie I tried to convert to `std::unique_ptr<>`, however `std::vector` doesn't accept subclass of `Geometry`in when pushing back e.g. a `std::unique_ptr<Cicle>` . How do you handle that ?

Reply:
@Hombre57 Eh? Works like a charm here...

```patch
diff --git a/rtgui/cropwindow.cc b/rtgui/cropwindow.cc
index 4b2c1a6cc..48bf5fad2 100644
--- a/rtgui/cropwindow.cc
+++ b/rtgui/cropwindow.cc
@@ -1838,23 +1838,23 @@ void CropWindow::expose (Cairo::RefPtr<Cairo::Context> cr)
                 }
 
                 // drawing Subscriber's visible geometry
-                const std::vector<Geometry*> visibleGeom = editSubscriber->getVisibleGeometry();
+                const auto& visibleGeom = editSubscriber->getVisibleGeometry();
                 cr->set_antialias(Cairo::ANTIALIAS_DEFAULT); // ANTIALIAS_SUBPIXEL ?
                 cr->set_line_cap(Cairo::LINE_CAP_SQUARE);
                 cr->set_line_join(Cairo::LINE_JOIN_ROUND);
 
                 // drawing outer lines
-                for (auto geom : visibleGeom) {
+                for (auto& geom : visibleGeom) {
                     geom->drawOuterGeometry(cr, this, *this);
                 }
 
                 // drawing inner lines
-                for (auto geom : visibleGeom) {
+                for (auto& geom : visibleGeom) {
                     geom->drawInnerGeometry(cr, this, *this);
                 }
 
                 // drawing to the "mouse over" channel
-                const auto mouseOverGeom = editSubscriber->getMouseOverGeometry();
+                const auto& mouseOverGeom = editSubscriber->getMouseOverGeometry();
                 if (mouseOverGeom.size()) {
                     if (mouseOverGeom.size() > 65534) {
                         // once it has been switched to OM_65535, it won't return back to OM_255
@@ -1876,7 +1876,7 @@ void CropWindow::expose (Cairo::RefPtr<Cairo::Context> cr)
                     crMO->fill();
 
                     int a=0;
-                    for (auto moGeom : mouseOverGeom) {
+                    for (auto& moGeom : mouseOverGeom) {
                         moGeom->drawToMOChannel(crMO, a, this, *this);
                         ++a;
                     }
diff --git a/rtgui/curveeditor.cc b/rtgui/curveeditor.cc
index fd41ba6e6..b5dd17182 100644
--- a/rtgui/curveeditor.cc
+++ b/rtgui/curveeditor.cc
@@ -449,7 +449,7 @@ void CurveEditor::switchOffEditMode ()
     EditSubscriber::switchOffEditMode();  // disconnect
 }
 
-const bool CurveEditor::mouseOver(const int modifierKey)
+bool CurveEditor::mouseOver(const int modifierKey)
 {
     EditDataProvider* provider = getEditProvider();
     subGroup->pipetteMouseOver(provider, modifierKey);
@@ -490,7 +490,7 @@ bool CurveEditor::drag1(const int modifierKey)
     return false;
 }
 
-const CursorShape CurveEditor::getCursor(const int objectID) const
+CursorShape CurveEditor::getCursor(const int objectID) const
 {
     if (remoteDrag) {
         return CSResizeHeight;
diff --git a/rtgui/curveeditor.h b/rtgui/curveeditor.h
index a4eeb0971..9d06ba22c 100644
--- a/rtgui/curveeditor.h
+++ b/rtgui/curveeditor.h
@@ -128,11 +128,11 @@ public:
     sigc::signal<void> signal_curvepoint_release();
 
     void switchOffEditMode () override;
-    const bool mouseOver(const int modifierKey) override;
+    bool mouseOver(int modifierKey) override;
     bool button1Pressed(const int modifierKey) override;
     bool button1Released() override;
     bool drag1(const int modifierKey) override;
-    const CursorShape getCursor(const int objectID) const override;
+    CursorShape getCursor(int objectID) const override;
 
 
 };
diff --git a/rtgui/editcallbacks.cc b/rtgui/editcallbacks.cc
index e130d7686..0a1b05abc 100644
--- a/rtgui/editcallbacks.cc
+++ b/rtgui/editcallbacks.cc
@@ -19,6 +19,8 @@
 
 #include "editcallbacks.h"
 
+#include "editwidgets.h"
+
 EditSubscriber::EditSubscriber (EditType editType) :
     ID(EUID_None),
     editingType(editType),
@@ -27,6 +29,8 @@ EditSubscriber::EditSubscriber (EditType editType) :
     action(EditSubscriber::Action::NONE)
 {}
 
+EditSubscriber::~EditSubscriber() = default;
+
 void EditSubscriber::setEditProvider(EditDataProvider *provider)
 {
     this->provider = provider;
@@ -38,7 +42,7 @@ void EditSubscriber::setEditID(EditUniqueID ID, BufferType buffType)
     bufferType = buffType;
 }
 
-const bool EditSubscriber::isCurrentSubscriber() const
+bool EditSubscriber::isCurrentSubscriber() const
 {
     //if (provider && provider->getCurrSubscriber())
     //  return provider->getCurrSubscriber()->getEditID() == ID;
@@ -69,27 +73,27 @@ void EditSubscriber::switchOffEditMode()
     unsubscribe();
 }
 
-const  EditUniqueID EditSubscriber::getEditID() const
+EditUniqueID EditSubscriber::getEditID() const
 {
     return ID;
 }
 
-const EditType EditSubscriber::getEditingType() const
+EditType EditSubscriber::getEditingType() const
 {
     return editingType;
 }
 
-const  BufferType EditSubscriber::getPipetteBufferType() const
+BufferType EditSubscriber::getPipetteBufferType() const
 {
     return bufferType;
 }
 
-const bool EditSubscriber::isDragging() const
+bool EditSubscriber::isDragging() const
 {
     return action == EditSubscriber::Action::DRAGGING;
 }
 
-const bool EditSubscriber::isPicking() const
+bool EditSubscriber::isPicking() const
 {
     return action == EditSubscriber::Action::PICKING;
 }
@@ -172,7 +176,7 @@ void EditDataProvider::setPipetteVal3(float newVal)
     pipetteVal3 = newVal;
 }
 
-const CursorShape EditDataProvider::getCursor(int objectID) const
+CursorShape EditDataProvider::getCursor(int objectID) const
 {
     if (currSubscriber) {
         currSubscriber->getCursor(objectID);
@@ -190,11 +194,11 @@ EditDataProvider* EditSubscriber::getEditProvider () {
     return provider;
 }
 
-const CursorShape EditSubscriber::getCursor (int objectID) const {
+CursorShape EditSubscriber::getCursor (int objectID) const {
     return CSHandOpen;
 }
 
-const bool EditSubscriber::mouseOver (const int modifierKey) {
+bool EditSubscriber::mouseOver (const int modifierKey) {
     return false;
 }
 
@@ -246,14 +250,16 @@ bool EditSubscriber::pick3 (const bool picked) {
     return false;
 }
 
-const std::vector<Geometry*>& EditSubscriber::getVisibleGeometry () {
+const std::vector<std::unique_ptr<Geometry>>& EditSubscriber::getVisibleGeometry()
+{
     return visibleGeometry;
 }
 
-const std::vector<Geometry*>& EditSubscriber::getMouseOverGeometry () {
+const std::vector<std::unique_ptr<Geometry>>& EditSubscriber::getMouseOverGeometry()
+{
     return mouseOverGeometry;
 }
 
-const int EditDataProvider::getPipetteRectSize () const {
+int EditDataProvider::getPipetteRectSize () const {
     return 8; // TODO: make a GUI
 }
diff --git a/rtgui/editcallbacks.h b/rtgui/editcallbacks.h
index 3ec15e00f..d11d261e4 100644
--- a/rtgui/editcallbacks.h
+++ b/rtgui/editcallbacks.h
@@ -61,19 +61,19 @@ public:
     void               setEditProvider(EditDataProvider *provider);
     EditDataProvider*  getEditProvider ();
     void               setEditID(EditUniqueID ID, BufferType buffType);
-    bool               isCurrentSubscriber();
+    bool               isCurrentSubscriber() const;
     virtual void       subscribe();
     virtual void       unsubscribe();
-    virtual void       switchOffEditMode ();    /// Occurs when the user want to stop the editing mode
-    EditUniqueID       getEditID();
-    EditType           getEditingType();
-    BufferType         getPipetteBufferType();
-    bool               isDragging();            /// Returns true if something is being dragged and drag events has to be sent (object mode only)
-    bool               isPicking();             /// Returns true if something is being picked
+    virtual void       switchOffEditMode();           /// Occurs when the user want to stop the editing mode
+    EditUniqueID       getEditID() const;
+    EditType           getEditingType() const;
+    BufferType         getPipetteBufferType() const;
+    bool               isDragging() const;            /// Returns true if something is being dragged and drag events has to be sent (object mode only)
+    bool               isPicking() const;             /// Returns true if something is being picked
 
     /** @brief Get the cursor to be displayed when above handles
     @param objectID object currently "hovered" */
-    virtual CursorShape getCursor (const int objectID);
+    virtual CursorShape getCursor (int objectID) const;
 
     /** @brief Triggered when the mouse is moving over an object
     This method is also triggered when the cursor is moving over the image in ET_PIPETTE mode
@@ -146,16 +146,10 @@ public:
     virtual bool pick3 (const bool picked);
 
     /** @brief Get the geometry to be shown to the user */
-    const std::vector<std::unique_ptr<Geometry>>& getVisibleGeometry()
-    {
-        return visibleGeometry;
-    }
+    const std::vector<std::unique_ptr<Geometry>>& getVisibleGeometry();
 
     /** @brief Get the geometry to be drawn in the "mouse over" channel, hidden from the user */
-    const std::vector<std::unique_ptr<Geometry>>& getMouseOverGeometry()
-    {
-        return mouseOverGeometry;
-    }
+    const std::vector<std::unique_ptr<Geometry>>& getMouseOverGeometry();
 };
 
 /** @brief Class to handle the furniture of data to the subscribers.
@@ -196,8 +190,8 @@ public:
     void setPipetteVal1(float newVal);
     void setPipetteVal2(float newVal);
     void setPipetteVal3(float newVal);
-    virtual const CursorShape getCursor(int objectID) const;
-    const int getPipetteRectSize () const;
+    virtual CursorShape getCursor(int objectID) const;
+    int getPipetteRectSize () const;
     EditSubscriber* getCurrSubscriber() const;
     virtual void getImageSize (int &w, int&h) = 0;
 };
diff --git a/rtgui/editwidgets.h b/rtgui/editwidgets.h
index ec935291e..3246d7377 100644
--- a/rtgui/editwidgets.h
+++ b/rtgui/editwidgets.h
@@ -538,5 +538,11 @@ inline Line::Line (int beginX, int beginY, int endX, int endY) :
         begin (beginX, beginY), end (endX, endY) {
 }
 
+#else
+
+class Geometry
+{
+};
+
 #endif
 
diff --git a/rtgui/gradient.cc b/rtgui/gradient.cc
index e4e006cbf..11a50b551 100644
--- a/rtgui/gradient.cc
+++ b/rtgui/gradient.cc
@@ -1,6 +1,8 @@
 /*
  *  This file is part of RawTherapee.
  */
+#include <array>
+
 #include "gradient.h"
 
 #include "editwidgets.h"
@@ -51,8 +53,6 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
     pack_start (*centerY, Gtk::PACK_SHRINK, 0);
 
     // Instantiating the Editing geometry; positions will be initialized later
-    Line *hLine, *vLine, *featherLine[2];
-    Circle *centerCircle;
 
     // Visible geometry
     std::unique_ptr<Line> hLine(new Line);
@@ -60,12 +60,11 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
     std::unique_ptr<Line> vLine(new Line);
     hLine->datum = vLine->datum = Geometry::IMAGE;
 
-    std::array<std::unique_ptr<Line>, 2> featherLine = {
-        new Line,
-        new Line
-    };
+    std::array<std::unique_ptr<Line>, 2> featherLine;
 
+    featherLine[0].reset(new Line);
     featherLine[0]->innerLineWidth = 2;
+    featherLine[1].reset(new Line);
     featherLine[1]->innerLineWidth = 2;
     featherLine[0]->datum = featherLine[1]->datum = Geometry::IMAGE;
 
@@ -87,12 +86,9 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
     vLine.reset(new Line);
     hLine->datum = vLine->datum = Geometry::IMAGE;
 
-    featherLine = {
-        new Line,
-        new Line
-    };
-
+    featherLine[0].reset(new Line);
     featherLine[0]->innerLineWidth = 2;
+    featherLine[1].reset(new Line);
     featherLine[1]->innerLineWidth = 2;
     featherLine[0]->datum = featherLine[1]->datum = Geometry::IMAGE;
 
@@ -113,13 +109,6 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
 
 Gradient::~Gradient()
 {
-    for (std::vector<Geometry*>::const_iterator i = visibleGeometry.begin(); i != visibleGeometry.end(); ++i) {
-        delete *i;
-    }
-
-    for (std::vector<Geometry*>::const_iterator i = mouseOverGeometry.begin(); i != mouseOverGeometry.end(); ++i) {
-        delete *i;
-    }
 }
 
 void Gradient::read (const ProcParams* pp, const ParamsEdited* pedited)
@@ -170,27 +159,27 @@ void Gradient::updateGeometry(const int centerX, const int centerY, const double
     const auto decay = feather * rtengine::norm2<double> (imW, imH) / 200.0;
     rtengine::Coord origin (imW / 2 + centerX * imW / 200, imH / 2 + centerY * imH / 200);
 
-    const auto updateLine = [&](Geometry* geometry, const float radius, const float begin, const float end)
+    const auto updateLine = [&](const std::unique_ptr<Geometry>& geometry, const float radius, const float begin, const float end)
     {
-        const auto line = static_cast<Line*>(geometry);
+        const auto line = static_cast<Line*>(geometry.get());
         line->begin = PolarCoord(radius, -degree + begin);
         line->begin += origin;
         line->end = PolarCoord(radius, -degree + end);
         line->end += origin;
     };
 
-    const auto updateLineWithDecay = [&](Geometry* geometry, const float radius, const float offSetAngle)
+    const auto updateLineWithDecay = [&](const std::unique_ptr<Geometry>& geometry, const float radius, const float offSetAngle)
     {
-        const auto line = static_cast<Line*>(geometry);
+        const auto line = static_cast<Line*>(geometry.get());
         line->begin = PolarCoord (radius, -degree + 180.) + PolarCoord (decay, -degree + offSetAngle);
         line->begin += origin;
         line->end = PolarCoord (radius, -degree) + PolarCoord (decay, -degree + offSetAngle);
         line->end += origin;
     };
 
-    const auto updateCircle = [&](Geometry* geometry)
+    const auto updateCircle = [&](const std::unique_ptr<Geometry>& geometry)
     {
-        const auto circle = static_cast<Circle*>(geometry);
+        const auto circle = static_cast<Circle*>(geometry.get());
         circle->center = origin;
     };
 
@@ -337,7 +326,7 @@ void Gradient::editToggled ()
     }
 }
 
-const CursorShape Gradient::getCursor(const int objectID) const
+CursorShape Gradient::getCursor(const int objectID) const
 {
     switch (objectID) {
     case (0):
@@ -363,7 +352,7 @@ const CursorShape Gradient::getCursor(const int objectID) const
     }
 }
 
-const bool Gradient::mouseOver(const int modifierKey)
+bool Gradient::mouseOver(const int modifierKey)
 {
     EditDataProvider* editProvider = getEditProvider();
 
diff --git a/rtgui/gradient.h b/rtgui/gradient.h
index 49799dc7b..8fbac6ece 100644
--- a/rtgui/gradient.h
+++ b/rtgui/gradient.h
@@ -52,8 +52,8 @@ public:
     void setEditProvider (EditDataProvider* provider) override;
 
     // EditSubscriber interface
-    const CursorShape getCursor(const int objectID) const override;
-    const bool mouseOver(const int modifierKey) override;
+    CursorShape getCursor(int objectID) const override;
+    bool mouseOver(int modifierKey) override;
     bool button1Pressed(const int modifierKey) override;
     bool button1Released() override;
     bool drag1(const int modifierKey) override;
```

Reply:
@Floessie Well... I clearly loose my time in this project, this seems so easy for you. Could you please commit this patch ?